### PR TITLE
Fix (most of) mem problems

### PR DESCRIPTION
**Pull request by ricrogz**
#### Reference Issue
This PR is in relation with issue #2107. It may be considered as a followup for PR #2113.

#### What does this implement/fix? Explain your changes.
This PR includes fixes to most mem problems detected by using what I suggested in #2113.

Most of the fixes are in test files, but there are also some fixes in RDKit code. Probably some of those fixes won't be the best solution (I tried not to break anything else), so please feel free to use those that you see fit, or to change them to something better.

Due to the extension of this PR, instead of submitting all of the changes in a single commit, I thought it might be a better idea to push my work branch directly, since each commit more or less addresses the problems discovered in a single test file.

An updated rdkit_valgrind.suppressios file is also provided. It contains suppressions for the following mem problems:

- Boost libraries (not really a problem, since the memory is recovered by the OS at program termination).
- Some leaks in externals MaeParser & Coordgen.
- Leaks in SMILES/SMARTS/SLN parsers. These leak on parser failures. I tried to address some of them by patching smiles.yy, but I am not familiar with flex/bison, so I cannot make much progress in that.

### CODE REVIEW

**Comment by greglandrum** [comment]
Thanks for this!
Since it's a big set of changes, it's going to take me a bit to go through them (and #2113) to make sure I'm comfortable with it. I will try to get it done this week.

**Comment by ricrogz** [comment]
No problem, Greg. It was actually fun chasing these.

Please let me know if you need any clarifications on the changes, or if you think there is a better option that the one I used (I'm sure there will be!).

File: Code/ForceField/MMFF/Params.cpp [hunk 1]
```diff
@@ -55,18 +52,15 @@ MMFFAromCollection::MMFFAromCollection(const boost::uint8_t *mmffArom) {
...
 extern const std::string defaultMMFFDef;
 
 MMFFDefCollection *MMFFDefCollection::getMMFFDef(const std::string &mmffDef) {
-  if (!ds_instance) {
-    ds_instance = new MMFFDefCollection(mmffDef);
-  } else if (!mmffDef.empty()) {
-    delete ds_instance;
-    ds_instance = new MMFFDefCollection(mmffDef);
+  if (!ds_instance || !mmffDef.empty()) {
+    ds_instance = std::unique_ptr<MMFFDefCollection>(new MMFFDefCollection(mmffDef));
```
Comment (greglandrum) [inline Code/ForceField/MMFF/Params.cpp]: Can't you just do this?
```suggestion
    ds_instance.reset(new MMFFDefCollection(mmffDef));
```

File: Code/ForceField/MMFF/Params.cpp [hunk 2]
```diff
@@ -449,19 +443,16 @@ const std::string defaultMMFFDef =
...
 
 MMFFPropCollection *MMFFPropCollection::getMMFFProp(
     const std::string &mmffProp) {
-  if (!ds_instance) {
-    ds_instance = new MMFFPropCollection(mmffProp);
-  } else if (!mmffProp.empty()) {
-    delete ds_instance;
-    ds_instance = new MMFFPropCollection(mmffProp);
+  if (!ds_instance || !mmffProp.empty()) {
+    ds_instance = std::unique_ptr<MMFFPropCollection>(new MMFFPropCollection(mmffProp));
```
Comment (greglandrum) [inline Code/ForceField/MMFF/Params.cpp]: same comment about using `.reset()` here

File: Code/ForceField/MMFF/testMMFFForceField.cpp [hunk 3]
```diff
@@ -1216,7 +1224,10 @@ int mmffValidationSuite(int argc, char *argv[]) {
                     ForceFields::MMFF::isDoubleZero(refTorsionEnergy))) {
                 fixTorsionInstance(torsionInstance);
                 refTorsionInstanceVec.push_back(torsionInstance);
-              }
+              } else {
+                delete torsionInstance;
+              torsionInstance;
```
Comment (greglandrum) [inline Code/ForceField/MMFF/testMMFFForceField.cpp]: Why is this here?

File: Code/ForceField/MMFF/testMMFFForceField.cpp [hunk 4]
```diff
@@ -1275,14 +1286,15 @@ int mmffValidationSuite(int argc, char *argv[]) {
...
+                  diff << "Expected:\n"
+                       << rdkLine << "\nFound:\n"
+                       << refLine << "\n";
                 }
                 errorMsg += diff.str();
               }
-              delete rdkTorsionInstanceVec[j];
-              delete refTorsionInstanceVec[j];
             }
+            BOOST_FOREACH (auto &ti, rdkTorsionInstanceVec) { delete ti; }
```
Comment (greglandrum) [inline Code/ForceField/MMFF/testMMFFForceField.cpp]: better to do this with:
```suggestion
            for (auto &ti : rdkTorsionInstanceVec) { delete ti; }
```

File: Code/GraphMol/FileParsers/SmilesMolSupplier.cpp [hunk 5]
```diff
@@ -41,6 +41,7 @@ SmilesMolSupplier::SmilesMolSupplier(const std::string &fileName,
   if (!tmpStream || (!(*tmpStream)) || (tmpStream->bad())) {
     std::ostringstream errout;
     errout << "Bad input file " << fileName;
+    if (tmpStream) { delete tmpStream; }
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/SmilesMolSupplier.cpp]: the `if` isn't necessary here; it's safe to call `delete` on `nullptr`

File: Code/GraphMol/FileParsers/TDTMolSupplier.cpp [hunk 6]
```diff
@@ -89,6 +89,7 @@ TDTMolSupplier::TDTMolSupplier(const std::string &fileName,
   if (!tmpStream || (!(*tmpStream)) || (tmpStream->bad())) {
     std::ostringstream errout;
     errout << "Bad input file " << fileName;
+    if (tmpStream) { delete tmpStream; }
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/TDTMolSupplier.cpp]: the `if` isn't necessary

File: Code/GraphMol/FilterCatalog/FilterCatalog.cpp [hunk 7]
```diff
@@ -58,7 +58,7 @@ bool FilterCatalogParams::addCatalog(FilterCatalogs catalog) {
   return addedCatalog;
 }
 
-void FilterCatalogParams::fillCatalog(FilterCatalog &catalog) {
+void FilterCatalogParams::fillCatalog(FilterCatalog &catalog) const {
```
Comment (greglandrum) [inline Code/GraphMol/FilterCatalog/FilterCatalog.cpp]: Nice... thanks for catching these additional `const` cases.

File: Code/GraphMol/FilterCatalog/FilterCatalog.h [hunk 8]
```diff
@@ -129,11 +129,12 @@ class RDKIT_FILTERCATALOG_EXPORT FilterCatalog : public FCatalog {
 
   FilterCatalog(FilterCatalogParams::FilterCatalogs catalogs)
       : FCatalog(), d_entries() {
-    setCatalogParams(new paramType_t(catalogs));
+    auto temp_params = static_cast<paramType_t>(catalogs);
```
Comment (greglandrum) [inline Code/GraphMol/FilterCatalog/FilterCatalog.h]: I shouldn't have been, but I was surprised that this works.
I still find it less than clear and would prefer:
```suggestion
    paramType_t temp_params(catalogs);
```

File: Code/GraphMol/Fingerprints/MACCS.cpp [hunk 9]
```diff
@@ -21,7 +21,8 @@
 #include <RDGeneral/BoostEndInclude.h>
 
 namespace {
-struct Patterns {
+class Patterns {
```
Comment (greglandrum) [inline Code/GraphMol/Fingerprints/MACCS.cpp]: seems like it would be more readable to have the members be `std::unique_ptr<RDKit::ROMol>`. Then this could stay a struct and you wouldn't need the dtor.

File: Code/GraphMol/SmilesParse/smarts.yy [hunk 10]
```diff
@@ -93,22 +95,43 @@ namespace {
 %left AND_TOKEN
 %right NOT_TOKEN
 
+%destructor { delete $$; } ATOM_TOKEN
```
Comment (greglandrum) [inline Code/GraphMol/SmilesParse/smarts.yy]: cool, I didn't know about this.

File: Code/GraphMol/new_canon.h [hunk 11]
```diff
@@ -96,13 +97,20 @@ struct RDKIT_GRAPHMOL_EXPORT canon_atom {
         isRingStereoAtom(false),
         nbrIds(NULL),
         p_symbol(NULL){};
+
+  ~canon_atom() {
+    if (nbrIds != nullptr) {
```
Comment (greglandrum) [inline Code/GraphMol/new_canon.h]: don't need the test for nullptr
Might think about making `nbrIds` an `std::unique_ptr`

**Comment by greglandrum** [comment]
I've now gone through all the changes to the non-testing code and think things mostly look good - just a few suggestions there.
I still want to confirm that everything builds and the tests pass on my windows machine, but I'm not anticipating any problems there.
With the questions/comments made above addressed, I'd be willing to merge this.

File: Code/ForceField/MMFF/Params.cpp (hunk 1 above)
Comment (ricrogz) [reply Code/ForceField/MMFF/Params.cpp]: Sure, I'll do.

File: Code/ForceField/MMFF/testMMFFForceField.cpp (hunk 3 above)
Comment (ricrogz) [reply Code/ForceField/MMFF/testMMFFForceField.cpp]: Oops... that "torsionInstance;" slipped in, somehow. I'll remove it.

File: Code/ForceField/MMFF/testMMFFForceField.cpp (hunk 4 above)
Comment (ricrogz) [reply Code/ForceField/MMFF/testMMFFForceField.cpp]: Ok.

File: Code/GraphMol/FileParsers/SmilesMolSupplier.cpp (hunk 5 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/SmilesMolSupplier.cpp]: Ok.

File: Code/GraphMol/FileParsers/TDTMolSupplier.cpp (hunk 6 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/TDTMolSupplier.cpp]: Ok.

File: Code/GraphMol/FilterCatalog/FilterCatalog.cpp (hunk 7 above)
Comment (ricrogz) [reply Code/GraphMol/FilterCatalog/FilterCatalog.cpp]: No problem. I think I needed to do it for the next comment to be successful.

File: Code/GraphMol/FilterCatalog/FilterCatalog.h (hunk 8 above)
Comment (ricrogz) [reply Code/GraphMol/FilterCatalog/FilterCatalog.h]: Ok.

File: Code/GraphMol/Fingerprints/MACCS.cpp (hunk 9 above)
Comment (ricrogz) [reply Code/GraphMol/Fingerprints/MACCS.cpp]: Ok, done.

File: Code/GraphMol/SmilesParse/smarts.yy (hunk 10 above)
Comment (ricrogz) [reply Code/GraphMol/SmilesParse/smarts.yy]: I found it in the docs, while trying to solve the problems with the parsers. I only managed it partially, and, since I am not too sure about what I tried, I did not update the *.tab.cpp.cmake files.

File: Code/GraphMol/new_canon.h (hunk 11 above)
Comment (ricrogz) [reply Code/GraphMol/new_canon.h]: Right.

But I think a `std::vector<int>`would be more appropriate than a `std::unique_ptr`.

File: Code/GraphMol/FileParsers/TDTMolSupplier.cpp [hunk 12]
```diff
@@ -89,7 +89,7 @@ TDTMolSupplier::TDTMolSupplier(const std::string &fileName,
   if (!tmpStream || (!(*tmpStream)) || (tmpStream->bad())) {
     std::ostringstream errout;
     errout << "Bad input file " << fileName;
-    if (tmpStream) { delete tmpStream; }
+    delete tmpStream;
```
Comment (bp-kelley) [inline Code/GraphMol/FileParsers/TDTMolSupplier.cpp]: I would point out that "delete" is now a code smell which is pretty nice.

One could do:
```
std::unique_ptr<std::istream*> tmpStream = ...
if (...) {
  std::ostringstream errout;
  errout << "Bad input file " << fileName;
  throw BadFileException(errout.str());
}
dp_inStream = tmpStream.release();
```

It sure is painful that we can't guarantee a move constructor for std::ifstream.  This problem would simply go away.

File: Code/GraphMol/FileParsers/SmilesMolSupplier.cpp [hunk 13]
```diff
@@ -41,7 +41,7 @@ SmilesMolSupplier::SmilesMolSupplier(const std::string &fileName,
   if (!tmpStream || (!(*tmpStream)) || (tmpStream->bad())) {
     std::ostringstream errout;
     errout << "Bad input file " << fileName;
-    if (tmpStream) { delete tmpStream; }
+    delete tmpStream;
```
Comment (bp-kelley) [inline Code/GraphMol/FileParsers/SmilesMolSupplier.cpp]: see comment for TDTMolSupplier

File: Code/GraphMol/FilterCatalog/FilterCatalog.h [hunk 14]
```diff
@@ -129,7 +129,7 @@ class RDKIT_FILTERCATALOG_EXPORT FilterCatalog : public FCatalog {
 
   FilterCatalog(FilterCatalogParams::FilterCatalogs catalogs)
       : FCatalog(), d_entries() {
-    auto temp_params = static_cast<paramType_t>(catalogs);
+    paramType_t temp_params(catalogs);
     setCatalogParams(&temp_params);
```
Comment (bp-kelley) [inline Code/GraphMol/FilterCatalog/FilterCatalog.h]: Taking an address of a temporary is a huge code smell here.  I realize that this is largely historical, but perhaps we should fix the API?

Comment (bp-kelley) [reply Code/GraphMol/FilterCatalog/FilterCatalog.h]:
```  virtual void setCatalogParams(const paramType &params) {
    // if we already have a paramter object throw an exception
    PRECONDITION(!dp_cParams,
                 "A parameter object already exists on the catalog");
    dp_cParams = new paramType(params);
  }
```

File: Code/GraphMol/new_canon.h [hunk 15]
```diff
@@ -99,9 +99,7 @@ class RDKIT_GRAPHMOL_EXPORT canon_atom {
         p_symbol(NULL){};
 
   ~canon_atom() {
-    if (nbrIds != nullptr) {
       free(nbrIds);
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/new_canon.h]: Why?

File: Code/GraphMol/FilterCatalog/FilterCatalog.h (hunk 14 above)
Comment (greglandrum) [reply Code/GraphMol/FilterCatalog/FilterCatalog.h]: I'd be ok with making that change, but not as part of this PR. If we're going to change the API it should be in a separate PR so that it's easy to see what was done.

Here it's worth adding a comment that we know that `setCatalogParams()` copies what is passed in.

**Comment by greglandrum** [comment]
@bp-kelley : this works for me on windows and linux and the travis tests are passing on both linux and the mac. I think it's fine to merge as is. You ok with that?

File: Code/GraphMol/new_canon.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (bp-kelley) [reply Code/GraphMol/new_canon.h]: You're right it's fine to free NULL.  I was misunderstanding the spec which is a bit embarrassing after all this time ;)

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/new_canon.h]: yeah, I was also really surprised when it was pointed out to me.

**Comment by bp-kelley** [comment]
Heck yes!  This is too much good work to ignore :). Merge away.

File: Code/GraphMol/FileParsers/MolFileParser.cpp [hunk 16]
```diff
@@ -956,7 +956,7 @@ void ParseNewAtomList(RWMol *mol, const std::string &text, unsigned int line) {
     }
 
     std::string atSymb = text.substr(pos, 4);
-    atSymb.erase(atSymb.find(" "), atSymb.size());
+    atSymb.erase(atSymb.find(' '), atSymb.size());
     int atNum = PeriodicTable::getTable()->getAtomicNumber(atSymb);
     if (!i) {
       a = new QueryAtom(*(mol->getAtomWithIdx(idx)));
```
[the remark this answers is not in the document]
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MolFileParser.cpp]: No, @d-b-w, as far as I understand it, the sentence in which `a` gets assigned a new pointer does only execute when for `i == 0`, so that only one pointed is created. Although I agree it is a bit confusing.
